include/triton/Tools/LinearLayout.h

```diff
@@ -0,0 +1,462 @@
+#ifndef TRITON_TOOLS_LINEARLAYOUT_H
+#define TRITON_TOOLS_LINEARLAYOUT_H
+
+#include <cstdint>
+#include <ostream>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "mlir/IR/BuiltinAttributes.h"
+#include "llvm/ADT/MapVector.h"
+#include "llvm/ADT/STLExtras.h"
+
+namespace mlir::triton {
+
+// # High-level overview of linear layouts
+//
+// The idea for linear layouts is due to Adam P. Goucher.
+//
+// In Triton, a linear layout (LL) is a function that maps from a "hardware
+// location" to a "logical tensor index".
+//
+// For example, suppose we have a 2D tensor T stored in GPU registers.  T's
+// layout is the function that, given a "hardware location" tuple of (thread-id,
+// warp-id), returns an index (x,y) into T.  In other words, if L(t,w) = (x,y)
```

Comment:
maybe we can also include `, cluster-cta-id` in the source tuple? This also has the nice added benefit that it makes it clear upfront that the arity of the source tuple may be different from that of the destination index.

Reply:
Although acknowledging it makes everything more bulky. So just a sentence to make it explicit that the arity doesn't have to be the same might be better

Reply:
Done